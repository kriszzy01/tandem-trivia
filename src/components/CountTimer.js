import React, { useLayoutEffect } from "react";
import { actionTypes } from "../utils";

export const CountTimer = ({ 
    answerIsCorrect, 
    dispatch, 
    timeLeft 
}) => {

    const fullTimeScale = 849.37;

    const getTimeRatio = () => (timeLeft / 10) * fullTimeScale - 0.1;

    useLayoutEffect(() => {
        const startTimer = setTimeout(() => {
            if (timeLeft !== 0 && answerIsCorrect === undefined) {
                dispatch({type: actionTypes.RUN_TIMER})
            }
        }, 2000);

        if (answerIsCorrect === undefined && timeLeft === 0) {
            dispatch({ type: actionTypes.SELECT_ANSWER, payload: false });
        }

        return () => {
            clearTimeout(startTimer);
        }
    }, [timeLeft, dispatch, answerIsCorrect]);

    return (
        <div className="timer">
            <svg version="1.1" className="timer__svg" viewBox="0 0 300 300" xmlns="http://www.w3.org/2000/svg" >
                <g fill="#fff" strokeLinecap="round" strokeWidth="25">
                    <circle cx="149.72" cy="153.06" r="135.25" stroke="rgb(9, 94, 67)" />
                    <path d="m284.97 153.06a135.25 135.25 0 0 1-135.25 135.25 135.25 135.25 0 0 1-135.25-135.25 135.25 135.25 0 0 1 135.25-135.25 135.25 135.25 0 0 1 135.25 135.25z"
                        stroke={timeLeft !== 0 ? "#1af6b3ff": "#e31227fc"} strokeDasharray={`${getTimeRatio()} ${fullTimeScale}`} />
                </g>
            </svg>
            <span className={`timer__label | font-base ${timeLeft === 0 ? "color-warning": "color-tertiary-shade"}`}>
                {timeLeft !== 0 ? timeLeft: "Time Up!"}
            </span>
        </div>
    );
};