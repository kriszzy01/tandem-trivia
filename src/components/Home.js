import React, { useState } from "react";
import {
    shuffleQuestions,
    actionTypes
} from "../utils";

export const Home = ({ dispatch }) => {
    const [rulesRead, setRulesRead] = useState(false);
    const [showAlert, setShowAlert] = useState(false);

    const handleBeginGame = () => {
        if (!rulesRead) {
            setShowAlert(true);

            return;
        }

        dispatch({
            type: actionTypes.START_GAME,
            payload: shuffleQuestions()
        })
    };

    const handleToggleCheckbox = () => {
        setRulesRead(!rulesRead);
        setShowAlert(false)
    };

    return (
        <section className="home">
            <header>
                <h1>Tandem Trivia</h1>
            </header>

            <div className="rules | gap-top-500">
                <h2>Game Rules</h2>
                <div className="gap-top-400">
                    <ul className="rules__list | flow">
                        <li>A Round has Ten(10) Random questions.</li>
                        <li>Each correct answer equals One(1) point.</li>
                        <li>Each question has a Time Limit of Ten(seconds).</li>
                        <li className="color-neutral">!--Answers SHUFFLE randomly every two seconds, so be CAREFULL and QUICK when selecting--!</li>
                    </ul>
                </div>
            </div>

            <div className="gap-top-500">
                <label htmlFor="rulesCheck">
                    <input
                        id="rulesCheck"
                        type="checkbox"
                        value="rulesRead"
                        onClick={handleToggleCheckbox}
                    />
                    <span>I'm Ready!</span>
                </label>
            </div>

            <button
                className="center | gap-top-500"
                onClick={handleBeginGame}
            >Start Game</button>

            {showAlert && <p className="center">Click the checkbox</p>}
        </section>
    );
};