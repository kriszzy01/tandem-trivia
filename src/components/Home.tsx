import React, { useState } from "react";
import { shuffleArray } from "../utils";
import { useGameState } from "../context";
import { Question } from "../types";
import triviaQuestions from "../assets/trivia_questions.json";

export const Home: React.FC = () => {
    const [rulesRead, setRulesRead] = useState(false);
    const [showAlert, setShowAlert] = useState(false);

    const [, dispatch] = useGameState();

    const handleBeginGame = (): void => {
        if (!rulesRead) {
            setShowAlert(true);

            return;
        }

        dispatch({
            type: "START_GAME",
            payload: shuffleArray<Question>(triviaQuestions)
        })
    };

    const handleToggleCheckbox = (): void => {
        setRulesRead(!rulesRead);
        setShowAlert(false)
    };

    return (
        <section className="home">
            <header>
                <h1>Tandem Trivia</h1>
            </header>

            <div className="rules | gap-top-500">
                <h2>Game Rules</h2>
                <div className="gap-top-400">
                    <ul className="rules__list | flow">
                        <li>A Round has Ten(10) Random questions.</li>
                        <li>Each correct answer equals One(1) point.</li>
                        <li>Each question has a Time Limit of Ten(seconds).</li>
                        <li className="color-neutral">!--Answers SHUFFLE randomly every two seconds, so be CAREFULL and QUICK when selecting--!</li>
                    </ul>
                </div>
            </div>

            <div className="gap-top-500">
                <label htmlFor="rulesCheck">
                    <input
                        id="rulesCheck"
                        type="checkbox"
                        value="rulesRead"
                        onClick={handleToggleCheckbox}
                    />
                    <span>I'm Ready!</span>
                </label>
            </div>

            <button
                className="center | gap-top-500"
                onClick={handleBeginGame}
            >Start Game</button>

            {showAlert && <p className="center">Click the checkbox</p>}
        </section>
    );
};