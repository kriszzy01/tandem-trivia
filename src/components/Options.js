import React, { useMemo } from "react";
import { actionTypes, shuffleQuestions } from "../utils";

export const Options = ({
    allOptions,
    correctOption,
    answerIsCorrect,
    dispatch,
    setShowResult,
    showResult,
    timeLeft }
) => {
    const submitAnswer = option => {
        if (option === correctOption) {
            dispatch({ type: actionTypes.SELECT_ANSWER, payload: true });
            setShowResult({...showResult, result: showResult.result + 1})
        } else {
            dispatch({ type: actionTypes.SELECT_ANSWER, payload: false });
        }
    }

    let shuffleOptions = useMemo(() => shuffleQuestions(allOptions), [allOptions]);

    return (
        <div className={`options`}>
            {shuffleOptions.map((option, index) =>
                <button
                    key={option}
                    onClick={() => submitAnswer(option)}
                    className={answerIsCorrect && option === correctOption ? "options__correct" : ""}
                    disabled={timeLeft < 1}>
                    <span>{option}</span>
                </button>
            )}
        </div>
    );
};