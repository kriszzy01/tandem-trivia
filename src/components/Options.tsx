import React, { useMemo } from "react";
import { shuffleArray } from "../utils";
import { useGameState } from "../context";
import { Question } from "../types";

interface OptionProps {
    allOptions: string[];
    correctOptions: string;
}

export const Options: React.FC<OptionProps> = ({ allOptions, correctOptions }) => {
    const [{ answerSelected, timeElapsed }, dispatch] = useGameState()

    const submitAnswer = (option: string): void => { //Omit Question
        if (option === correctOptions) {
            dispatch({ type: "SELECT_ANSWER", payload: true });
            //setShowResult({...showResult, result: showResult.result + 1})
        } else {
            dispatch({ type: "SELECT_ANSWER", payload: false });
        }
    }

    let shuffleOptions = useMemo(() => shuffleArray<string>(allOptions), [allOptions]);

    return (
        <div className={`options`}>
            {shuffleOptions.map((option, index) =>
                <button
                    key={`${Date.now()} ${index}`}
                    onClick={() => submitAnswer(option)}
                    className={answerSelected && option === correctOptions ? "options__correct" : ""}
                    disabled={!Boolean(timeElapsed) || answerSelected}>
                    <span>{option}</span>
                </button>
            )}
        </div>
    );
};