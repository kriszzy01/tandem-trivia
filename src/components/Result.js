import React from "react";
import { actionTypes, shuffleQuestions } from "../utils";

export const Result = ({ result, dispatch }) => {
    const handleNewRound = () => {
        dispatch({ type: actionTypes.REPLAY_GAME, payload: shuffleQuestions() })
    }

    return (
        <div className="result center">
            <ul className="result__list">
                <h1>Your Score: {result}/10</h1>
                <li>Incorrect: {`${10 - result}`}</li>
                <li>Correct: {result}</li>
            </ul>

            <button onClick={handleNewRound}>New Round</button>
        </div>
    );
};