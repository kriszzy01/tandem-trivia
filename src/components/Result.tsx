import React from "react";
import { shuffleArray } from "../utils";
import { useGameState } from "../context";

export const Result: React.FC<{result: number}> = ({ result }) => {
    const [, dispatch] = useGameState();

    const handleNewRound = () => {
        //dispatch({ type: actionTypes.REPLAY_GAME, payload: shuffleQuestions() })
    }

    return (
        <div className="result center">
            <ul className="result__list">
                <h1>Your Score: {result}/10</h1>
                <li>Incorrect: {`${10 - result}`}</li>
                <li>Correct: {result}</li>
            </ul>

            <button onClick={handleNewRound}>New Round</button>
        </div>
    );
};