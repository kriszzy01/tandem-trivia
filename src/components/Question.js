import React, { useState } from "react";
import { Options } from "./Options";
import { Remark } from "./Remark";
import { CountTimer } from "./CountTimer";
import { Result } from "./Result";

export const Question = ({
    newRound,
    questionNumber,
    gameStatus,
    answerIsCorrect,
    timeLeft,
    dispatch
}) => {
    const [showResult, setShowResult] = useState({
        result: 0,
        isResultPage: false,
    })

    const selectedQuestion = newRound[questionNumber - 1];

    const incorrectOptions = selectedQuestion.incorrect;

    const correctOptions = selectedQuestion.correct;

    const allOptions = incorrectOptions.concat(correctOptions);

    return (
        <>
            {!showResult.isResultPage ?
                <section className="question | flow">
                    <div className="question__header | splitter">
                        <h2 className="center">{`Question ${questionNumber}/${newRound.length}`}</h2>
                        <CountTimer
                            answerIsCorrect={answerIsCorrect}
                            timeLeft={timeLeft}
                            dispatch={dispatch} />
                    </div>

                    <div className="question__main">
                        <svg width="366.86" height="245.49" version="1.1" viewBox="0 0 91.714 61.373" xmlns="http://www.w3.org/2000/svg">
                            <g fill="#fff">
                                <g transform="matrix(.92162 0 0 .91569 -2.9371 -10.988)">
                                    <path d="m8.552 32.501c1.9338-4.4356 5.3737-8.1111 9.377-10.829 4.0033-2.7182 8.5601-4.5292 13.206-5.881 17.821-5.1847 37.796-3.5443 53.886 5.7052 4.1676 2.3957 8.1674 5.3903 10.592 9.541 1.7457 2.9885 2.5898 6.472 2.5683 9.933-0.02146 3.4609-0.89274 6.8958-2.3761 10.023-2.9667 6.2541-8.2792 11.154-14.184 14.766-13.29 8.1283-29.947 10.159-45.071 6.4243-8.962-2.2131-17.68-6.5583-23.486-13.735-2.9029-3.5884-5.0211-7.8449-5.8922-12.378-0.87117-4.5327-0.46476-9.3383 1.3798-13.569z" fill="#fff" />
                                </g>
                                <ellipse cx="76.883" cy="52.245" rx="5.2077" ry="4.9838" strokeLinecap="round" strokeWidth="7.1216" />
                                <ellipse cx="84.947" cy="58.405" rx="3.4158" ry="2.8559" strokeLinecap="round" strokeWidth="4.3661" />
                            </g>
                        </svg>
                        <p className="imposter | color-tertiary-shade">{selectedQuestion.question}</p>
                    </div>

                    <Options
                        allOptions={allOptions}
                        correctOption={correctOptions}
                        dispatch={dispatch}
                        timeLeft={timeLeft}
                        answerIsCorrect={answerIsCorrect}
                        setShowResult={setShowResult}
                        showResult={showResult}
                    />

                    <Remark
                        dispatch={dispatch}
                        gameStatus={gameStatus}
                        questionNumber={questionNumber}
                        answerIsCorrect={answerIsCorrect}
                        correctOption={correctOptions}
                        setShowResult={setShowResult}
                        showResult={showResult}
                    />
                </section> :
                <Result 
                    result={showResult.result} 
                    dispatch={dispatch}/>
            }
        </>
    );
};