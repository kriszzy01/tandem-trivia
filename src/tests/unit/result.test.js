import React from "react";
import "@testing-library/jest-dom/extend-expect";
import { screen, render } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { Result } from "../../components/Result";

const resultProps = {
    result: 6,
    dispatch: jest.fn()
};

describe("Result", () => {
    test("it renders all components", () => {
        const container = render(<Result {...resultProps} />);

        expect(container).toMatchSnapshot();
    });

    test("it calls new round function", () => {
        render(<Result {...resultProps} />);

        userEvent.click(screen.getByRole("button"));

        expect(resultProps.dispatch).toHaveBeenCalled();
    });
});