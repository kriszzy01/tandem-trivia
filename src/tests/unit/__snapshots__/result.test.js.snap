// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Result it renders all components 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="result center"
      >
        <ul
          class="result__list"
        >
          <h1>
            Your Score: 
            6
            /10
          </h1>
          <li>
            Incorrect: 
            4
          </li>
          <li>
            Correct: 
            6
          </li>
        </ul>
        <button>
          New Round
        </button>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="result center"
    >
      <ul
        class="result__list"
      >
        <h1>
          Your Score: 
          6
          /10
        </h1>
        <li>
          Incorrect: 
          4
        </li>
        <li>
          Correct: 
          6
        </li>
      </ul>
      <button>
        New Round
      </button>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
